File: tests/test_smooth_voxel.py
#!/usr/bin/env python3
"""
Test the smooth voxel grid function
"""
import numpy as np
import evlib

# Use wrapper function instead of direct access
create_smooth_voxel_grid = evlib.create_smooth_voxel_grid


def test_events_to_smooth_voxel_grid_py():
    """Test creating a smooth voxel grid from events"""
    # Create sample event data with non-integer positions for interpolation test
    xs = np.array([10, 20, 30, 40], dtype=np.int64)
    ys = np.array([15, 25, 35, 45], dtype=np.int64)
    ts = np.array([0.1, 0.3, 0.6, 0.9], dtype=np.float64)
    ps = np.array([1, -1, 1, -1], dtype=np.int64)

    num_bins = 5
    resolution = (50, 50)  # (width, height)

    # Create voxel grid with default (trilinear) interpolation
    voxel_grid = create_smooth_voxel_grid(xs, ys, ts, ps, num_bins, resolution)

    # Check shape and type
    assert voxel_grid.shape == (num_bins, resolution[1], resolution[0])
    assert voxel_grid.dtype == np.float32

    # Check that the voxel grid contains non-zero elements
    assert np.sum(voxel_grid) > 0


def test_events_to_smooth_voxel_grid_py_empty():
    """Test creating a smooth voxel grid from empty events"""
    # Create empty event data
    xs = np.array([], dtype=np.int64)
    ys = np.array([], dtype=np.int64)
    ts = np.array([], dtype=np.float64)
    ps = np.array([], dtype=np.int64)

    num_bins = 3
    resolution = (50, 50)  # (width, height)

    # Create voxel grid
    voxel_grid = create_smooth_voxel_grid(xs, ys, ts, ps, num_bins, resolution)

    # Check shape and type
    assert voxel_grid.shape == (num_bins, resolution[1], resolution[0])
    assert voxel_grid.dtype == np.float32

    # Check that the voxel grid is all zeros
    assert np.all(voxel_grid == 0)


def test_different_interpolation_methods():
    """Test different interpolation methods for smooth voxel grid"""
    # Create sample event data
    xs = np.array([10, 20, 30, 40], dtype=np.int64)
    ys = np.array([15, 25, 35, 45], dtype=np.int64)
    ts = np.array([0.1, 0.3, 0.6, 0.9], dtype=np.float64)
    ps = np.array([1, -1, 1, -1], dtype=np.int64)

    num_bins = 5
    resolution = (50, 50)  # (width, height)

    # Test different interpolation methods
    interpolation_methods = ["trilinear", "bilinear", "temporal"]

    for method in interpolation_methods:
        # Create smooth voxel grid
        voxel_grid = create_smooth_voxel_grid(xs, ys, ts, ps, num_bins, resolution, method)

        # Check shape and type
        assert voxel_grid.shape == (num_bins, resolution[1], resolution[0])
        assert voxel_grid.dtype == np.float32

        # Sum of all voxel values should be non-zero
        assert np.sum(voxel_grid) != 0


if __name__ == "__main__":
    # Run the tests manually if this file is executed
    test_events_to_smooth_voxel_grid_py()
    test_events_to_smooth_voxel_grid_py_empty()
    test_different_interpolation_methods()
    print("All tests passed!")

File: tests/test_smooth_voxel.rs
//! Tests for smooth voxel grid representation

#[cfg(test)]
mod tests {
    use evlib::ev_core::{Event, Events};
    use evlib::ev_representations::smooth_voxel_grid::{
        events_to_smooth_voxel_grid, events_to_trilinear_voxel_grid, SmoothEventsToVoxelGrid,
    };

    /// Create some test events for use in tests
    fn create_test_events() -> Events {
        let mut events = Events::with_capacity(5);
        events.push(Event {
            x: 10,
            y: 15,
            t: 0.1,
            polarity: 1,
        });
        events.push(Event {
            x: 20,
            y: 25,
            t: 0.2,
            polarity: -1,
        });
        events.push(Event {
            x: 30,
            y: 35,
            t: 0.3,
            polarity: 1,
        });
        events.push(Event {
            x: 40,
            y: 45,
            t: 0.4,
            polarity: -1,
        });
        events.push(Event {
            x: 50,
            y: 55,
            t: 0.5,
            polarity: 1,
        });
        events
    }

    #[test]
    fn test_smooth_voxel_grid_constructor() {
        let converter = SmoothEventsToVoxelGrid::new(5, 100, 100, Some("trilinear".to_string()));
        assert_eq!(converter.num_bins, 5);
        assert_eq!(converter.width, 100);
        assert_eq!(converter.height, 100);
        assert_eq!(converter.interpolation, "trilinear");

        // Test default interpolation
        let converter = SmoothEventsToVoxelGrid::new(5, 100, 100, None);
        assert_eq!(converter.interpolation, "trilinear");
    }

    #[test]
    fn test_empty_events() {
        let events = Events::new();
        let result = events_to_smooth_voxel_grid(&events, 5, 100, 100, None).unwrap();
        let shape = result.shape();

        assert_eq!(shape.dims().len(), 3);
        assert_eq!(shape.dims()[0], 5);
        assert_eq!(shape.dims()[1], 100);
        assert_eq!(shape.dims()[2], 100);

        // Verify all values are zero
        let data = result.to_vec1().unwrap();
        assert!(data.iter().all(|&x| x == 0.0));
    }

    #[test]
    fn test_trilinear_interpolation() {
        let events = create_test_events();
        let result =
            events_to_smooth_voxel_grid(&events, 5, 100, 100, Some("trilinear".to_string()))
                .unwrap();
        let shape = result.shape();

        assert_eq!(shape.dims().len(), 3);
        assert_eq!(shape.dims()[0], 5);
        assert_eq!(shape.dims()[1], 100);
        assert_eq!(shape.dims()[2], 100);

        // Verify the grid has non-zero values
        let data = result.to_vec1().unwrap();
        assert!(data.iter().any(|&x| x != 0.0));
    }

    #[test]
    fn test_bilinear_interpolation() {
        let events = create_test_events();
        let result =
            events_to_smooth_voxel_grid(&events, 5, 100, 100, Some("bilinear".to_string()))
                .unwrap();
        let shape = result.shape();

        assert_eq!(shape.dims().len(), 3);
        assert_eq!(shape.dims()[0], 5);
        assert_eq!(shape.dims()[1], 100);
        assert_eq!(shape.dims()[2], 100);

        // Verify the grid has non-zero values
        let data = result.to_vec1().unwrap();
        assert!(data.iter().any(|&x| x != 0.0));
    }

    #[test]
    fn test_temporal_interpolation() {
        let events = create_test_events();
        let result =
            events_to_smooth_voxel_grid(&events, 5, 100, 100, Some("temporal".to_string()))
                .unwrap();
        let shape = result.shape();

        assert_eq!(shape.dims().len(), 3);
        assert_eq!(shape.dims()[0], 5);
        assert_eq!(shape.dims()[1], 100);
        assert_eq!(shape.dims()[2], 100);

        // Verify the grid has non-zero values
        let data = result.to_vec1().unwrap();
        assert!(data.iter().any(|&x| x != 0.0));
    }

    #[test]
    fn test_convenience_function() {
        let events = create_test_events();
        let result = events_to_trilinear_voxel_grid(&events, 5, 100, 100).unwrap();
        let shape = result.shape();

        assert_eq!(shape.dims().len(), 3);
        assert_eq!(shape.dims()[0], 5);
        assert_eq!(shape.dims()[1], 100);
        assert_eq!(shape.dims()[2], 100);

        // Verify the grid has non-zero values
        let data = result.to_vec1().unwrap();
        assert!(data.iter().any(|&x| x != 0.0));
    }
}
